Confusing diagnostics for unstable features when building Rust with stable rustc

I'm trying to build Rust on a Tier 3 platform (`riscv64-alpine-linux-musl`; actually it's `riscv64-unknown-musl` that's listed in Tier 3, but I don't think that affects the problem). Since there are no snapshots for this platform, I installed rustc from the distro's package manager (apk) and created a `config.toml` file to force using the pre-installed rustc rather than downloading snapshots:

```
profile = "compiler"

[build]
rustc  = "/usr/bin/rustc"
cargo  = "/usr/bin/cargo"
rustfmt = "/usr/bin/rustfmt"
host   = ["riscv64-alpine-linux-musl"]
target = ["riscv64-alpine-linux-musl"]

[rust]
musl-root = "/usr"
debug     = true

[target.riscv64-alpine-linux-musl]
musl-libdir = "/usr/lib"
```

And the output of `rustc --version -v`:

```
# rustc --version -v
rustc 1.68.2 (9eb3afe9e 2023-03-27) (Alpine Linux)
binary: rustc
commit-hash: 9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0
commit-date: 2023-03-27
host: riscv64-alpine-linux-musl
release: 1.68.2
LLVM version: 15.0.7
```

I ran `./x.py build`; when it got as far as building the core library, I got a series of errors that were very hard to make sense of until I realized that it's because core library code depends on unstable features and I was using a stable rustc to compile it:

```
building stage0 library artifacts (riscv64-alpine-linux-musl) 
  Downloaded adler v1.0.2
  Downloaded miniz_oxide v0.5.3
  Downloaded unicode-width v0.1.10
  Downloaded hashbrown v0.12.3
  Downloaded compiler_builtins v0.1.91
  Downloaded cc v1.0.77
  Downloaded addr2line v0.17.0
  Downloaded rustc-demangle v0.1.21
  Downloaded gimli v0.26.2
  Downloaded 9 crates (1.2 MB) in 1.45s
   Compiling cc v1.0.77
   Compiling compiler_builtins v0.1.91
   Compiling core v0.0.0 (/home/scratch/rust/library/core)
error: expected identifier, found keyword `move`
   --> library/core/src/ops/try_trait.rs:395:15
    |
395 |         const move |a, b| NeverShortCircuit(f(a, b))
    |               ^^^^ expected identifier, found keyword

error: expected one of `:`, `;`, or `=`, found `|`
   --> library/core/src/ops/try_trait.rs:395:20
    |
395 |         const move |a, b| NeverShortCircuit(f(a, b))
    |                    ^ expected one of `:`, `;`, or `=`
```

There were many more errors, but I think they all relate to unstable features.

It took me ages to figure out that (for example) the first error is due to the `const_closures` feature being disabled in stable.

I understand that the stable compiler won't allow using unstable features even with `-Zunstable-options`, but is it possible for the parser to recognize this code even so and reject it with something like "feature `const_closures` is not enabled" instead of the much more generic "expected identifier" error?

Ah ok, it was related to the context of the `const` closure: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9c71507b2e0273d5062102e2cc0756e9

If you look at the same gist with nightly, the error message is much better:
```
error[E0658]: const trait impls are experimental
 --> src/lib.rs:5:24
  |
5 | const fn foo() -> impl ~const FnMut(usize, usize) -> NeverShortCircuit {
  |                        ^^^^^^
  |
  = note: see issue #67792 <https://github.com/rust-lang/rust/issues/67792> for more information
  = help: add `#![feature(const_trait_impl)]` to the crate attributes to enable

error[E0658]: const closures are experimental
 --> src/lib.rs:6:5
  |
6 |     const move |a, b| NeverShortCircuit(f(a, b))
  |     ^^^^^
  |
  = note: see issue #106003 <https://github.com/rust-lang/rust/issues/106003> for more information
  = help: add `#![feature(const_closures)]` to the crate attributes to enable
```
So I think this particular diagnostic issue has already been fixed. I still want to keep this issue open to track checking the version of the bootstrap compiler, though.

--- a/src/bootstrap/Cargo.toml
+++ b/src/bootstrap/Cargo.toml
@@ -57,6 +57,7 @@ walkdir = "2"
 sysinfo = { version = "0.26.0", optional = true }
 clap = { version = "4.2.4", default-features = false, features = ["std", "usage", "help", "derive", "error-context"] }
 clap_complete = "4.2.2"
+semver = "1.0.17"
 
 # Solaris doesn't support flock() and thus fd-lock is not option now
 [target.'cfg(not(target_os = "solaris"))'.dependencies]

--- a/src/bootstrap/config.rs
+++ b/src/bootstrap/config.rs
@@ -24,6 +24,7 @@ pub use crate::flags::Subcommand;
 use crate::flags::{Color, Flags, Warnings};
 use crate::util::{exe, output, t};
 use once_cell::sync::OnceCell;
+use semver::Version;
 use serde::{Deserialize, Deserializer};
 use serde_derive::Deserialize;
 
@@ -1019,6 +1020,8 @@ impl Config {
             config.download_beta_toolchain();
             config.out.join(config.build.triple).join("stage0/bin/rustc")
         });
+        config.check_build_rustc_version();
+
         config.initial_cargo = build
             .cargo
             .map(|cargo| {
@@ -1678,6 +1681,37 @@ impl Config {
 
     pub fn default_codegen_backend(&self) -> Option<Interned<String>> {
         self.rust_codegen_backends.get(0).cloned()
+    }
+
+    fn check_build_rustc_version(&self) {
+        // check rustc version is same or lower with 1 apart from the building one
+        let mut cmd = Command::new(&self.initial_rustc);
+        cmd.arg("--version");
+        let rustc_output = output(&mut cmd)
+            .lines()
+            .next()
+            .unwrap()
+            .split(' ')
+            .nth(1)
+            .unwrap()
+            .split('-')
+            .next()
+            .unwrap()
+            .to_owned();
+        let rustc_version = Version::parse(&rustc_output.trim()).unwrap();
+        let source_version =
+            Version::parse(&fs::read_to_string(self.src.join("src/version")).unwrap().trim())
+                .unwrap();
+        if !(source_version == rustc_version
+            || (source_version.major == rustc_version.major
+                && source_version.minor == rustc_version.minor + 1))
+        {
+            let prev_version = format!("{}.{}.x", source_version.major, source_version.minor - 1);
+            panic!(
+                "Unexpected rustc version: {}, we should use {}/{} to build source with {}",
+                rustc_version, prev_version, source_version, source_version
+            );
+        }
     }
 
     /// Returns the commit to download, or `None` if we shouldn't download CI artifacts.

--- a/src/bootstrap/config.rs
+++ b/src/bootstrap/config.rs
@@ -1710,10 +1710,11 @@ impl Config {
                 && source_version.minor == rustc_version.minor + 1))
         {
             let prev_version = format!("{}.{}.x", source_version.major, source_version.minor - 1);
-            panic!(
+            eprintln!(
                 "Unexpected rustc version: {}, we should use {}/{} to build source with {}",
                 rustc_version, prev_version, source_version, source_version
             );
+            crate::detail_exit(1);
         }
     }
 

--- a/src/bootstrap/config.rs
+++ b/src/bootstrap/config.rs
@@ -1020,7 +1020,6 @@ impl Config {
             config.download_beta_toolchain();
             config.out.join(config.build.triple).join("stage0/bin/rustc")
         });
-        config.check_build_rustc_version();
 
         config.initial_cargo = build
             .cargo
@@ -1683,7 +1682,7 @@ impl Config {
         self.rust_codegen_backends.get(0).cloned()
     }
 
-    fn check_build_rustc_version(&self) {
+    pub fn check_build_rustc_version(&self) {
         if self.dry_run() {
             return;
         }

--- a/src/bootstrap/config.rs
+++ b/src/bootstrap/config.rs
@@ -1684,6 +1684,10 @@ impl Config {
     }
 
     fn check_build_rustc_version(&self) {
+        if self.dry_run() {
+            return;
+        }
+
         // check rustc version is same or lower with 1 apart from the building one
         let mut cmd = Command::new(&self.initial_rustc);
         cmd.arg("--version");

--- a/src/bootstrap/lib.rs
+++ b/src/bootstrap/lib.rs
@@ -414,6 +414,7 @@ impl Build {
                 bootstrap_out.display()
             )
         }
+        config.check_build_rustc_version();
 
         if rust_info.is_from_tarball() && config.description.is_none() {
             config.description = Some("built from a source tarball".to_owned());

